### PR TITLE
[ISSUE-2409]add trino engineconn-plugins

### PR DESCRIPTION
**Pull request by ws00428637**
### What is the purpose of the change
EngineConn-Plugin for Trino
Related issues: #2409.

### Brief change log
- Define an executor base on Trino jdbc client;
- Add progress support for Trino on jdbc plugin

### Verifying this change
This change is a trivial rework / code cleanup without any test coverage.  

### Does this pull request potentially affect one of the following parts:
- Dependencies (does it add or upgrade a dependency): yes
- Anything that affects deployment: don't know
- The MGS(Microservice Governance Services), i.e., Spring Cloud Gateway, OpenFeign, Eureka.: no

### Documentation
- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? not documented

### CODE REVIEW

File: linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>${httpclient.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpmime</artifactId>
+            <version>${httpmime.version}</version>
+        </dependency>
+        <dependency>
```
Comment (wForget) [inline linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml:72]: useless dependency?

File: linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            <resource>
+                <directory>src/main/resources</directory>
+                <excludes>
+                    <exclude>**/application.yml</exclude>
+                    <exclude>**/bootstrap.yml</exclude>
+                </excludes>
+            </resource>
+        </resources>
+    </build>
+</project>
```
Comment (wForget) [inline linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml:149]: newline

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/java/org/apache/linkis/engineplugin/trino/protocol/GatewayPasswordCallback.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package org.apache.linkis.engineplugin.trino.protocol;
```
Comment (wForget) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/java/org/apache/linkis/engineplugin/trino/protocol/GatewayPasswordCallback.java:1]: license

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/resources/linkis-engineconn.properties [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#
+
+#
+wds.linkis.server.version=v1
+wds.linkis.engineconn.debug.enable=true
+#wds.linkis.keytab.enable=true
+wds.linkis.engineconn.plugin.default.class=org.apache.linkis.engineplugin.trino.TrinoEngineConnPlugin
+
+wds.linkis.engine.trino.url=jdbc:trino://192.168.1.61:20080/hive/default
+#wds.linkis.engine.trino.url=jdbc:trino://dap-trino.weizhipin.com:443/hive/default
```
Comment (wForget) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/resources/linkis-engineconn.properties:23]: use 127.0.0.1

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/java/org/apache/linkis/engineplugin/trino/protocol/StaticPasswordCallback.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package org.apache.linkis.engineplugin.trino.protocol;
```
Comment (wForget) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/java/org/apache/linkis/engineplugin/trino/protocol/StaticPasswordCallback.java:1]: license

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/exception/NotSupportedTrinoTypeException.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+  *
+  * @param erroCode 41004
+  * @param desc hive query fail
+  */
+case class TrinoQueryFailedException(erroCode:Int,
+                                    desc:String) extends ErrorException(erroCode, desc){
+
+}
```
Comment (wForget) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/exception/NotSupportedTrinoTypeException.scala:41]: newline

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/executor/TrinoEngineConnExecutor.scala [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+import java.util.Properties
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.atomic.AtomicBoolean
+import javax.security.auth.callback.PasswordCallback
+import scala.collection.JavaConversions._
+import scala.collection.mutable.{ArrayBuffer, ListBuffer}
+
+class TrinoEngineConnExecutor(id: Int, url: String, properties: Properties, passwordCallback: PasswordCallback) extends ComputationExecutor with Logging {
+
+  private val log = LoggerFactory.getLogger(getClass)
```
Comment (wForget) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/executor/TrinoEngineConnExecutor.scala:50]: used `with Logging`, no need to create log object.

File: linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  ~ limitations under the License.
+  -->
+  
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>linkis</artifactId>
+        <groupId>org.apache.linkis</groupId>
+        <version>1.0.3</version>
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml:25]: should update to 1.1.3

File: linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.linkis</groupId>
+            <artifactId>linkis-common</artifactId>
+            <version>${linkis.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.linkis</groupId>
+            <artifactId>linkis-message-scheduler</artifactId>
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml:58]: should rename linkis-rpc

File: linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+            <version>${httpclient.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpmime</artifactId>
+            <version>${httpmime.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml:74]: lombok is not recommended

File: linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.linkis</groupId>
+            <artifactId>linkis-message-scheduler</artifactId>
+            <version>${linkis.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/pom.xml:64]: Does trino depend on httpclient?

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/java/org/apache/linkis/engineplugin/trino/protocol/GatewayPasswordCallback.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.apache.commons.io.IOUtils;
+
+import javax.security.auth.callback.PasswordCallback;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * @author knypys 2022/4/15 11:40
+ */
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/java/org/apache/linkis/engineplugin/trino/protocol/GatewayPasswordCallback.java:11]: author and date should be removed

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/resources/linkis-engineconn.properties [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+#
+wds.linkis.server.version=v1
+wds.linkis.engineconn.debug.enable=true
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/resources/linkis-engineconn.properties:18]: debug can be removed

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/resources/log4j2-engineconn.xml [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+  ~ 
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+  
+<configuration status="error" monitorInterval="30">
+    <appenders>
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/resources/log4j2-engineconn.xml:20]: log4j can refer to the configuration of the shell engine plugin

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/TrinoEngineConnPlugin.scala [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+  private var engineLaunchBuilder: EngineConnLaunchBuilder = _
+
+  private var engineFactory: EngineConnFactory = _
+
+  private val defaultLabels: List[Label[_]] = new util.ArrayList[Label[_]]()
+
+  override def init(params: util.Map[String, Any]): Unit = {
+    val typeMap = new util.HashMap[String,String]()
+    typeMap.put("type","trino")
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/TrinoEngineConnPlugin.scala:45]: can be removed

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/TrinoEngineConnPlugin.scala [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    typeMap.put("version","371")
+    val typeLabel =new EngineTypeLabel()
+    typeLabel.setValue(typeMap)
+    this.defaultLabels.add(typeLabel)
+  }
+
+  override def getEngineResourceFactory(): EngineResourceFactory = {
+    EP_CONTEXT_CONSTRUCTOR_LOCK.synchronized{
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/TrinoEngineConnPlugin.scala:53]: can refer to the shell engine

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/common/TrinoUtils.scala [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+package org.apache.linkis.engineplugin.trino.common
+
+/**
+ * @author knypys 2022/4/15 13:04
+ */
+object TrinoUtils {
+
+  def msDurationToString(ms: Long): String = {
+    val second = 1000
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/common/TrinoUtils.scala:25]: use ByteTimeUtils.msDurationToString method

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/conf/TrinoEngineConfiguration.scala [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+ 
+package org.apache.linkis.engineplugin.trino.conf
+
+import org.apache.linkis.common.conf.CommonVars
+
+/**
+ * @author knypys 2022/4/15 13:21
+ */
+object TrinoEngineConfiguration {
+  val TRINO_URL: CommonVars[String] = CommonVars("wds.linkis.engine.trino.url", "jdbc:trino://127.0.0.1:8080/hive/default")
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/conf/TrinoEngineConfiguration.scala:26]: should remove wds prefix

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/creation/TrinoEngineConnFactory.scala [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * @author knypys 2022/4/15 13:47
+ */
+class TrinoEngineConnFactory extends ComputationSingleExecutorEngineConnFactory with Logging {
+
+  override protected def newExecutor(id: Int, engineCreationContext: EngineCreationContext, engineConn: EngineConn): LabelExecutor = {
+    engineConn.getEngineConnSession match {
+      case trinoSession: TrinoSession =>
+        new TrinoEngineConnExecutor(id, trinoSession.url, trinoSession.properties, trinoSession.passwordCallback)
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/creation/TrinoEngineConnFactory.scala:46]: You can refer to the presto engine implementation, which should be able to support tasks to adjust different trino clusters

File: linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/executor/TrinoEngineConnExecutor.scala [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+import java.util
+import java.util.Properties
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.atomic.AtomicBoolean
+import javax.security.auth.callback.PasswordCallback
+import scala.collection.JavaConversions._
+import scala.collection.mutable.{ArrayBuffer, ListBuffer}
+
+class TrinoEngineConnExecutor(id: Int, url: String, properties: Properties, passwordCallback: PasswordCallback) extends ComputationExecutor with Logging {
+
```
Comment (peacewong) [inline linkis-engineconn-plugins/engineconn-plugins/trino/src/main/scala/org/apache/linkis/engineplugin/trino/executor/TrinoEngineConnExecutor.scala:49]: should extends ConcurrentComputationExecutor

**Review by peacewong — APPROVED** [review]

**Pull request closed without merging**